### ISSUE TITLE
lego dns-challenge not inputting correct gcloud domain

### ISSUE THREAD

**Issue by jaredhendrickson13** [go-acme/lego#874]
I recently tore down and am rebuilding a new server for certificate management. I use Google Cloud for DNS. When I check my available DNS zones from gcloud command line tool I find the correct domains listed:

```
$ gcloud dns managed-zones list
NAME                     DNS_NAME                  DESCRIPTION                                     VISIBILITY
jh-dot-co            jh.co.                The official DNS short name for internal use.
jaredhendrickson-dot-com  jaredhendrickson.com.      Our main jaredhendrickson.com DNS records.
```

Everything works perfectly for my shortname jh.co. However, when I run my lego command for my full jaredhendrickson.com domain name, I get an error:

 ```
$ GCE_PROJECT="jaredhendrickson-dns" GCE_DOMAIN="new.jaredhendrickson.com" lego --email="Jared@jaredhendrickson.com" --domains="new.jaredhendrickson.com" --dns="gcloud" --path="/etc/ssl/certs/new.jaredhendrickson.com/" --server=https://acme-v02.api.letsencrypt.org/directory --dns.resolvers="10.1.1.2:53" run
2019/05/01 11:24:07 Please review the TOS at https://letsencrypt.org/documents/LE-SA-v1.2-November-15-2017.pdf
2019/05/01 11:24:07 Do you accept the TOS? Y/n
y
2019/05/01 11:24:09 [INFO][new.jaredhendrickson.com] acme: Obtaining bundled SAN certificate
2019/05/01 11:24:10 [INFO][new.jaredhendrickson.com] AuthURL: https://acme-v02.api.letsencrypt.org/acme/authz/<scrubbed>
2019/05/01 11:24:10 [INFO][new.jaredhendrickson.com] acme: Trying to solve DNS-01
2019/05/01 11:24:10 Could not obtain certificates
	acme: Error -> One or more domains had a problem:
[new.jaredhendrickson.com] Error presenting token: No matching GoogleCloud domain found for domain com.
```

The error I get is very interesting to me, because if I run the same command with a domain name I do not manage (e.g. example.com), it specifies `No matching GoogleCloud domain found for domain example.com` whereas with my domain, it states `No matching GoogleCloud domain found for domain com.` and does not seem to use the full domain name I am passing into the command during the DNS challenge. 

Any help is greatly appreciated!

**Comment by mhoran** [comment]
Hey @jaredhendrickson13, based on the error message, it looks like you're running a pretty "old" version of `lego`. There have been quite a few changes to the DNS providers since then, and this issue may have been fixed.

If simply upgrading to the latest version doesn't fix the issue, try setting `GCE_DEBUG=true` for more verbose logging. Also, so far as I can tell, `GCE_DOMAIN` doesn't do anything outside of the tests, so I think you can drop that.

**Comment by jaredhendrickson13** [comment]
Looks like you are correct, my setup script was installing lego v1.2.1. However, updating to lego v2.5.0 still does not resolve the previous problem for my primary domain name, and my DNS shortname is now throwing a different error when I run the command. 

Shortname:
```
$ GCE_PROJECT="jaredhendrickson-dns" GCE_DEBUG=true lego --email="Jared@jaredhendrickson.com" --domains="new.jh.co" --dns="gcloud" --path="/etc/ssl/certs/new.jh.co" --server=https://acme-v02.api.letsencrypt.org/directory --dns.resolvers="10.1.1.2:53" run
2019/05/02 08:08:18 [INFO] [new.jh.co] acme: use dns-01 solver
2019/05/02 08:08:18 [INFO] [new.jh.co] acme: Preparing to solve DNS-01
2019/05/02 08:08:19 [INFO] [new.jh.co] acme: Cleaning DNS-01 challenge
2019/05/02 08:08:20 [WARN] [new.jh.co] acme: error cleaning up: googlecloud: no public zone found for domain jh.co.
2019/05/02 08:08:20 Could not obtain certificates:
	acme: Error -> One or more domains had a problem:
[new.jh.co] [new.jh.co] acme: error presenting token: googlecloud: no public zone found for domain jh.co.
```

I have verified in GoogleCloud that the DNS zone does exist and is a public zone. Also odd that it did not seem to have this problem on lego v1.2.1 as certificates were generating fine for the jh.co domain name. 

Primary name:
```
$ GCE_PROJECT="jaredhendrickson-dns" GCE_DEBUG=true lego --email="Jared@jaredhendrickson.com" --domains="new.jaredhendrickson.com" --dns="gcloud" --path="/etc/ssl/certs/new.jaredhendrickson.com/" --server=https://acme-v02.api.letsencrypt.org/directory --dns.resolvers="10.1.1.2:53" run
2019/05/02 08:23:13 [INFO] [new.jaredhendrickson.com] acme: use dns-01 solver
2019/05/02 08:23:13 [INFO] [new.jaredhendrickson.com] acme: Preparing to solve DNS-01
2019/05/02 08:23:14 [INFO] [new.jaredhendrickson.com] acme: Cleaning DNS-01 challenge
2019/05/02 08:23:14 [WARN] [new.jaredhendrickson.com] acme: error cleaning up: googlecloud: no matching domain found for domain com.
2019/05/02 08:23:14 Could not obtain certificates:
	acme: Error -> One or more domains had a problem:
[new.jaredhendrickson.com] [new.jaredhendrickson.com] acme: error presenting token: googlecloud: no matching domain found for domain com.
```

This appears to behave the same as it was on lego v1.2.1. I did discover if I change the --dns.resolvers value to point straight to GoogleCloud's nameservers rather than our internal nameserver, I get the same error as jh.co:

```
$ GCE_PROJECT="jaredhendrickson-dns" GCE_DEBUG="true" lego --email="Jared@jaredhendrickson.com" --domains="new.jaredhendrickson.com" --dns="gcloud" --path="/etc/ssl/certs/new.jaredhendrickson.com/" --server=https://acme-v02.api.letsencrypt.org/directory --dns.resolvers="ns-cloud-c1.googledomains.com:53" --dns.resolvers="ns-cloud-c2.googledomains.com:53" --dns.resolvers="ns-cloud-c3.googledomains.com:53" --dns.resolvers="ns-cloud-c4.googledomains.com:53" run
2019/05/02 08:59:16 [WARN] [new.jaredhendrickson.com] acme: error cleaning up: googlecloud: no public zone found for domain jaredhendrickson.com.
2019/05/02 08:59:17 Could not obtain certificates:
	acme: Error -> One or more domains had a problem:
[new.jaredhendrickson.com] [new.jaredhendrickson.com] acme: error presenting token: googlecloud: no public zone found for domain jaredhendrickson.com.
```

After setting GCE_DEBUG=true, the only new items I am seeing in my logs at `/home/certmgr/.config/gcloud/logs` are:

```
2019-05-02 08:06:51,637 DEBUG    root            No implementation for release track [GA] for element: [/sm/google-cloud-sdk/lib/surface/web_security_scanner]
2019-05-02 08:06:51,638 DEBUG    root            No implementation for release track [GA] for element: [/sm/google-cloud-sdk/lib/surface/tasks]
2019-05-02 08:06:51,643 DEBUG    root            Loaded Command Group: [u'gcloud', u'feedback']
2019-05-02 08:06:51,645 DEBUG    root            No implementation for release track [GA] for element: [/sm/google-cloud-sdk/lib/surface/storage]
2019-05-02 08:06:51,647 DEBUG    root            Loaded Command Group: [u'gcloud', u'app']
2019-05-02 08:06:51,650 DEBUG    root            No implementation for release track [GA] for element: [/sm/google-cloud-sdk/lib/surface/scc]
2019-05-02 08:06:51,651 DEBUG    root            No implementation for release track [GA] for element: [/sm/google-cloud-sdk/lib/surface/data_catalog]
2019-05-02 08:06:51,652 DEBUG    root            Loaded Command Group: [u'gcloud', u'composer']
2019-05-02 08:06:51,670 DEBUG    root            Loaded Command Group: [u'gcloud', u'ai_platform']
2019-05-02 08:06:51,673 DEBUG    root            Loaded Command Group: [u'gcloud', u'iam']
2019-05-02 08:06:51,675 DEBUG    root            No implementation for release track [GA] for element: [/sm/google-cloud-sdk/lib/surface/genomics]
2019-05-02 08:06:51,676 DEBUG    root            No implementation for release track [GA] for element: [/sm/google-cloud-sdk/lib/surface/billing]
2019-05-02 08:06:51,687 DEBUG    root            Loaded Command Group: [u'gcloud', u'help']
2019-05-02 08:06:51,690 DEBUG    root            Loaded Command Group: [u'gcloud', u'source']
2019-05-02 08:06:51,694 DEBUG    root            Loaded Command Group: [u'gcloud', u'init']
2019-05-02 08:06:51,696 DEBUG    root            Loaded Command Group: [u'gcloud', u'version']
2019-05-02 08:06:51,697 DEBUG    root            Loaded Command Group: [u'gcloud', u'asset']
2019-05-02 08:06:51,700 DEBUG    root            Loaded Command Group: [u'gcloud', u'dns']
2019-05-02 08:06:51,702 DEBUG    root            No implementation for release track [GA] for element: [/sm/google-cloud-sdk/lib/surface/iap]
2019-05-02 08:06:51,704 DEBUG    root            No implementation for release track [GA] for element: [/sm/google-cloud-sdk/lib/surface/resources]
2019-05-02 08:06:51,705 DEBUG    root            Loaded Command Group: [u'gcloud', u'pubsub']
2019-05-02 08:06:51,708 DEBUG    root            Loaded Command Group: [u'gcloud', u'info']
2019-05-02 08:06:51,711 DEBUG    root            Loaded Command Group: [u'gcloud', u'organizations']
2019-05-02 08:06:51,713 DEBUG    root            No implementation for release track [GA] for element: [/sm/google-cloud-sdk/lib/surface/emulators]
2019-05-02 08:06:51,714 DEBUG    root            Loaded Command Group: [u'gcloud', u'auth']
2019-05-02 08:06:51,716 DEBUG    root            No implementation for release track [GA] for element: [/sm/google-cloud-sdk/lib/surface/search_help]
2019-05-02 08:06:51,717 DEBUG    root            No implementation for release track [GA] for element: [/sm/google-cloud-sdk/lib/surface/scheduler]
2019-05-02 08:06:51,718 DEBUG    root            Loaded Command Group: [u'gcloud', u'spanner']
2019-05-02 08:06:51,720 DEBUG    root            No implementation for release track [GA] for element: [/sm/google-cloud-sdk/lib/surface/recommender]
2019-05-02 08:06:51,721 DEBUG    root            Loaded Command Group: [u'gcloud', u'projects']
2019-05-02 08:06:51,723 DEBUG    root            Loaded Command Group: [u'gcloud', u'resource_manager']
2019-05-02 08:06:51,725 DEBUG    root            Loaded Command Group: [u'gcloud', u'topic']
2019-05-02 08:06:51,742 DEBUG    root            Loaded Command Group: [u'gcloud', u'compute']
2019-05-02 08:06:51,746 DEBUG    root            Loaded Command Group: [u'gcloud', u'builds']
2019-05-02 08:06:51,749 DEBUG    root            Loaded Command Group: [u'gcloud', u'dataflow']
2019-05-02 08:06:51,751 DEBUG    root            Loaded Command Group: [u'gcloud', u'ml']
2019-05-02 08:06:51,753 DEBUG    root            No implementation for release track [GA] for element: [/sm/google-cloud-sdk/lib/surface/error_reporting]
2019-05-02 08:06:51,756 DEBUG    root            Loaded Command Group: [u'gcloud', u'debug']
2019-05-02 08:06:51,758 DEBUG    root            Loaded Command Group: [u'gcloud', u'endpoints']
2019-05-02 08:06:51,760 DEBUG    root            Loaded Command Group: [u'gcloud', u'deployment_manager']
2019-05-02 08:06:51,761 DEBUG    root            Loaded Command Group: [u'gcloud', u'access_context_manager']
2019-05-02 08:06:51,763 DEBUG    root            Loaded Command Group: [u'gcloud', u'meta']
2019-05-02 08:06:51,765 DEBUG    root            Loaded Command Group: [u'gcloud', u'datastore']
2019-05-02 08:06:51,769 DEBUG    root            Loaded Command Group: [u'gcloud', u'functions']
2019-05-02 08:06:51,772 DEBUG    root            No implementation for release track [GA] for element: [/sm/google-cloud-sdk/lib/surface/monitoring]
2019-05-02 08:06:51,772 DEBUG    root            Loaded Command Group: [u'gcloud', u'redis']
2019-05-02 08:06:51,774 DEBUG    root            No implementation for release track [GA] for element: [/sm/google-cloud-sdk/lib/surface/dialogflow]
2019-05-02 08:06:51,776 DEBUG    root            No implementation for release track [GA] for element: [/sm/google-cloud-sdk/lib/surface/dlp]
2019-05-02 08:06:51,777 DEBUG    root            Loaded Command Group: [u'gcloud', u'kms']
2019-05-02 08:06:51,779 DEBUG    root            Loaded Command Group: [u'gcloud', u'config']
2019-05-02 08:06:51,781 DEBUG    root            No implementation for release track [GA] for element: [/sm/google-cloud-sdk/lib/surface/cloud_shell]
2019-05-02 08:06:51,782 DEBUG    root            No implementation for release track [GA] for element: [/sm/google-cloud-sdk/lib/surface/run]
2019-05-02 08:06:51,783 DEBUG    root            Loaded Command Group: [u'gcloud', u'dataproc']
2019-05-02 08:06:51,785 DEBUG    root            Loaded Command Group: [u'gcloud', u'ml_engine']
2019-05-02 08:06:51,786 DEBUG    root            Loaded Command Group: [u'gcloud', u'iot']
2019-05-02 08:06:51,788 DEBUG    root            Loaded Command Group: [u'gcloud', u'firebase']
2019-05-02 08:06:51,790 DEBUG    root            Loaded Command Group: [u'gcloud', u'filestore']
2019-05-02 08:06:51,792 DEBUG    root            No implementation for release track [GA] for element: [/sm/google-cloud-sdk/lib/surface/bq]
2019-05-02 08:06:51,793 DEBUG    root            Loaded Command Group: [u'gcloud', u'sql']
2019-05-02 08:06:51,795 DEBUG    root            Loaded Command Group: [u'gcloud', u'services']
2019-05-02 08:06:51,803 DEBUG    root            Loaded Command Group: [u'gcloud', u'container']
2019-05-02 08:06:51,806 DEBUG    root            Loaded Command Group: [u'gcloud', u'bigtable']
2019-05-02 08:06:51,808 DEBUG    root            Loaded Command Group: [u'gcloud', u'logging']
2019-05-02 08:06:51,810 DEBUG    root            Loaded Command Group: [u'gcloud', u'service_management']
2019-05-02 08:06:51,812 DEBUG    root            No implementation for release track [GA] for element: [/sm/google-cloud-sdk/lib/surface/remote_build_execution]
2019-05-02 08:06:51,815 DEBUG    root            No implementation for release track [GA] for element: [/sm/google-cloud-sdk/lib/surface/runtime_config]
2019-05-02 08:06:51,823 DEBUG    root            No implementation for release track [GA] for element: [/sm/google-cloud-sdk/lib/surface/survey]
2019-05-02 08:06:51,823 DEBUG    root            Loaded Command Group: [u'gcloud', u'components']
2019-05-02 08:06:51,825 DEBUG    root            No implementation for release track [GA] for element: [/sm/google-cloud-sdk/lib/surface/healthcare]
2019-05-02 08:06:51,826 DEBUG    root            Loaded Command Group: [u'gcloud', u'domains']
2019-05-02 08:06:51,830 DEBUG    root            Loaded Command Group: [u'gcloud', u'docker']
2019-05-02 08:06:51,942 DEBUG    root            No implementation for release track [GA] for element: [/sm/google-cloud-sdk/lib/surface/interactive]
2019-05-02 08:06:51,943 DEBUG    root            No implementation for release track [GA] for element: [/sm/google-cloud-sdk/lib/surface/firestore]
```

Nothing in the logs seem particularly relevant to issue, unless I am looking at the wrong logs? I have installed google-cloud-sdk manually, and have installed them via apt and snap with the same result. This problem persists on both Ubuntu 16.04 and Ubuntu 18.04.

**Comment by mhoran** [comment]
Okay, I was able to reproduce `no matching domain found for domain com`, and I think this is related to #352. I have an existing zone `test.matthoran.com`. I added `not-matt.com`. I tried to request a certificate for `test.not-matt.com` and got `no matching domain found for domain com`.

Note that `not-matt.com` is not a real domain, so I just want to make sure that you have set up delegation properly from `jaredhendrickson.com` and `jh.co` to rule that out.

I was also able to reproduce `no public zone found for domain` from another GCE account, but I've been unable to reproduce it in mine. There are lots of funky things about that particular account, so there could be something else going on there.

**Comment by mhoran** [comment]
Upon further investigation, this does seem to be a delegation issue -- especially given the behavior you see when you use the Cloud DNS servers as your resolver. In `FindZoneByFqdnCustom` we search up to the first SOA record that is returned for the given domain. In the case of `test.not-matt.com`, given `not-matt.com` doesn't exist, the first existing SOA record is for `com`. That gets handed back to the gcloud DNS provider, and Google returns 0 managed zones corresponding to `com`.

I'm not quite sure about the private zone issue -- do you have any private zones in the account? The code is pretty straightforward, but it does seem to imply that there are private zones being returned by Cloud DNS, and none of them match the domain returned by `FindZOneByFqdnCustom`.

Regardless, I think a good workaround for both this issue and #352 would be to allow users to override the auto-selected zone as is possible with the route53 provider.

**Comment by jaredhendrickson13** [comment]
Thank you sir, this did help me find the cause for the `no matching domain` errors. The nameservers I was querying were not returning the SOA for each domain correctly. Our internal DNS server would find the SOA for jh.co but not jaredhendrickson.com and Google Cloud's nameservers were doing the opposite. I found that 8.8.8.8 returns the SOA for both domains correctly and switching the --dns.resolvers value to 8.8.8.8 cleared up that issue. With these changes, reverting back to lego v1.2.1 gets things working for me for now. 

However, I still encounter the `no public zone found` error on lego v2.5.0. This is odd as lego v1.2.1 finds the zones and successfully completes the challenge without issue. I've verified that all our Google Cloud DNS zones are set to public and that the gcloud command line tools are still able to see the managed zones. From my knowledge, we have never set any zones to private in Google Cloud, only public.
